flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -117,7 +117,7 @@ public void increaseResourceRequirementsBy(ResourceCounter increment) {
 
     @Override
     public void decreaseResourceRequirementsBy(ResourceCounter decrement) {
-        if (decrement.isEmpty()) {
+        if (decrement.isEmpty() || totalResourceRequirements.isEmpty()) {
```

Comment:
The new branch would hide a bug. There should never be a case where we decrease the resource requirement such that the result is negative. If such a case exists we should try to fix that instead.

Reply:
IIUC, the negative requirement is by design not allowed. Could @zhuzhurk make sure about this?
The `ResourceCount` used here already covers this case.

Reply:
I guess what @zentol means is that if we decrease a non-empty `ResourceCounter` when the `totalResourceRequirements` is not empty, it is an unexpected bug and needs to be exposed.
So I think we should add a check here and also fix `ResourceCounter#updateNewValue()` to throw exception if negative result happens.

Reply:
Ahh. I think I misunderstand it.
> There should never be a case where we decrease the resource requirement such that the result is negative.
This scenario already exists, that is why we observe multiple "Clearing resource requirements of job". The `totalResourceRequirements` will decrease each time in `onReleaseTaskManager`. If that is a bug, we should fix it. @zentol